### DISCUSSION TITLE
Python: Build out API to be on-par with "dataframe" libraries

### DISCUSSION THREAD

**Discussion by binste** [duckdb/duckdb#12134] [Ideas]
I believe DuckDB has most of what it needs be used in the place of a “dataframe” library like Pandas or Polars. However, it requires Python users to write SQL which means writing code in strings which can be tricky to craft and you risk SQL injections, it lacks code highlighting, etc. The [Ibis](https://ibis-project.org/) API around DuckDB provides a way to overcome this. DuckDB itself has some APIs which already get you quite far as well but they can't really compete yet with what Polars or Pandas have to offer. Some logic can only be expressed in SQL and not through the Python APIs.

Polars has:
*  [contexts](https://docs.pola.rs/user-guide/concepts/contexts/) which is very similar to the [Relational API](https://duckdb.org/docs/api/python/relational_api) in DuckDB
*  [expressions](https://docs.pola.rs/user-guide/concepts/expressions/) which are a significantly more powerful version of DuckDB's [Expression API](https://duckdb.org/docs/api/python/expression)

What are your thoughts on extending these APIs and the documentation to provide a pythonic alternative to Polars, Pandas, etc.? Maybe you already have a clear vision of where you want to go (Spark API?) or this is not a priority. Fully understand both, just wanted to drop these ideas in case it's helpful to you and because I really like the library! :) If you're open to contributions on this, happy to help with Python code and writing documentation!

Some related discussions I found: https://github.com/duckdb/duckdb-web/issues/1859, https://github.com/duckdb/duckdb-web/issues/1151, and https://github.com/duckdb/duckdb/discussions/8242

# Some first ideas
Setup for examples below
```python
import duckdb
from duckdb import ColumnExpression, ConstantExpression, FunctionExpression, Expression
```

## `FunctionExpression` with code completion
```python
class _SingleFunc:
    def __init__(self, name: str, doc: str | None) -> None:
        self.name = name
        self.__doc__ = doc
        self.__name__ = name

    def __call__(self, *args) -> FunctionExpression:
        parsed_args: list[Expression] = []
        for arg in args:
            if isinstance(arg, Expression):
                parsed_args.append(arg)
            else:
                parsed_args.append(ConstantExpression(arg))

        return FunctionExpression(self.name, *parsed_args)


class _Func:
    def __init__(self) -> None:
        # TODO: Would be better to autogenerate the methods so that we can include
        # arguments as well and docstrings in a way that also work for static code
        # analysis tools in non-interactive environments
        self._functions: dict[str : str | None] = {
            r[0]: r[-1]
            for r in duckdb.sql(
                """SELECT DISTINCT ON(function_name)
            function_name,
            function_type,
            return_type,
            parameters,
            parameter_types,
            description
        FROM duckdb_functions()
        WHERE function_type = 'scalar'"""
            ).fetchall()
        }

    def __getattr__(self, name: str):
        if name in self._functions:
            return _SingleFunc(name, doc=self._functions[name])
        else:
            return super().__getattr__(name)

    def __dir__(self) -> list[str]:
        return list(self._functions)


func = _Func()
```

Now we have a more convenient way of creating `FunctionExpression`s with code completion:

<img width="594" alt="image" src="https://github.com/duckdb/duckdb/assets/23366411/5169c808-da8b-450d-a776-81eaef7b6c7a">

and docstrings:

<img width="1027" alt="image" src="https://github.com/duckdb/duckdb/assets/23366411/69f2f80c-152f-443a-b2a5-67804661c8e4">

It would probably be better to autogenerate the methods on `_Func` instead of using `__getattr__` and `__dir__` so that this also works in non-interactive environments and with static code analysis tools.

## Simplify creating `ColumnExpression`
Simplest way would be just to alias `ColumnExpression`:
```python
col = ColumnExpression
```
Would be in line with Polars `pl.col` which works fine.

## Example
Putting the above 2 together would give:

```python
import pandas

input_df = pandas.DataFrame.from_dict(
    {"i": [1, 2, 3, 4], "j": ["one", "two", "three", "four"]}
)

rel = duckdb.from_df(input_df)

# Modified version of https://duckdb.org/docs/guides/python/relational_api_pandas. Currently possible:
transformed_rel = (
    rel.filter("ends_with(j,'e')")
    .project("i, j, i*2 as two_i")
    .order("i desc")
    .limit(2)
)

# Rewrite with new functionality:
transformed_rel = (
    rel.filter(func.ends_with(col("j"), "e"))
    .project(
        col("i"),
        # Or stick to strings for simple column selections
        "j",
        (col("i") * 2).alias("two_i"),
    )
    # Order does not work but sort does
    .sort(col("i").desc())
    .limit(2)
)
```

## Columns on `DuckDBPyRelation`
I think it could be even more convenient if we could get column expressions by accessing columns with square bracket notation on a `DuckDBPyRelation`, e.g. `rel["column_name"]`. We could get autocompletion for column names with this class method on `DuckDBPyRelation`:
```python
# If this method would be added to DuckDBPyRelation, it would give autocompletion
# for column names in square brackets, e.g. rel[<TAB>]
def _ipython_key_completions_(self):
    return self.columns
```

`_SingleFunc.__call__` could be extended to also handle this kind of column type so you could do:

```python
transformed_rel_2 = (
    rel.filter(func.ends_with(rel["j"], "e"))
    .project(
        rel["i"],
        # Or stick to strings for simple column selections
        "j",
        (rel["i"] * 2).alias("two_i"),
    )
    # Order does not work but sort does
    .sort(rel["i"].desc())
    .limit(2)
)
```
We could use the alias of `rel` in joins to declare join columns.

## Column selectors
Column selectors for `ColumnExpression` like in Polars (and Ibis). Duckdb already has for example `select_dtypes` but that’s not an expression.

## Window functions
Support window functions in expressions. I think this was done for R in https://github.com/duckdb/duckdb/pull/6757

Can do this in the Relational API but that’s not returning an expression -> Can't name the resulting column: `rel.row_number("over (partition by a)", projected_columns="a")`.

**Comment by cmdlineluser** [comment]
In case it is of any use: when I was previously looking for info on this topic I found: 

https://github.com/duckdb/duckdb/issues/2000#issuecomment-1849927294

> We discussed this internally and concluded to close this, focusing our efforts on the PySpark-compatible API now.

https://duckdb.org/docs/api/python/spark_api

**Reply by binste** [reply]
Thanks for linking to that discussion! 

The Spark API builds on the Expression and Relational APIs and I understand that discussion as being about if it's worth it to build a Pandas-compatible API and that they'd rather build out the PySpark-compatible API instead.

However, I think that features such as window functions would need to be introduced in the Expression and Relational APIs first anyway and I'd see a great adoption potential for the later two APIs as there is a clear mapping to DuckDB SQL (function names are the same, etc.). Curious to hear more about this :)

**Reply by Tishj** [reply]
I agree with this, Expressions are great and I'd love to see the implementation be extended and used more.
That being said, it's also really easy to make contributions here, it's more typing intensive than it actually requires thinking through problems.

What happens for the non-expression API is we take a string, convert that to ParsedExpressions and those then get bound+executed.
The only step the expression API changes is the step "string -> ParsedExpression".

Instead we use a dedicated python method to create an Expression object, which internally contains a ParsedExpression.
When the Expression is provided to the system we just take the ParsedExpression from the Expression, no further processing is impacted by this.

The biggest hurdle is often the API design, as not to impact the existing methods, but still make the method that takes the Expression feel nice to interact with.

**Reply by binste** [reply]
Thank you for the context @Tishj! Regarding the suggestions in the above post, do you see anything where you'd specifically appreciate a contribution/appreciate if I explore this further? For me, from a user perspective, `FunctionExpression with code completion` and `Simplify creating ColumnExpression` would be very handy and don't seem to mess with the current API design as it just produces expressions which can be again used by the relational API as-is.

**Reply by Tishj** [reply]
Yea that sounds good, I had an idea for this before
We have .json files defining the prototypes of a lot of our functions, used for generating boilerplate, I think that json could also be used for this purpose.

**Reply by binste** [reply]
Found the files. I got it working in https://github.com/binste/duckdb/tree/autogenerate_python_functions. Based on the contributing guidelines I understand that you prefer not to have draft PRs so I'll create a PR once I've found some more time to thoroughly test it and clean up the code.

**Reply by Tishj** [reply]
That's looking great already!

**Reply by Tishj** [reply]
I see that you're analyzing the `parameters` string, we might want to add these assertions to the `tools/pythonpkg/scripts/generate_functions_code.py` script so we can verify that they stay in tact

`parameters` aren't really used for anything other than just a constant string in the generated struct currently, it makes sense to analyze it to generate the python code but none of that is enforced currently

The plan is to move the pythonpkg out of core to its own repo relatively soon, at which point it will be even harder to make sure this stays in tact, so we definitely need to add these assertions to the generate script in core 👍

**Reply by binste** [reply]
Just fyi, I'm moving apartments next week so I'll probably get back to this in ~2 weeks. Looking forward to wrapping it up 👍🏼

**Comment by binste** [comment]
Another idea would be to add a `.plot` interface. See [Pandas](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.plot.html) which uses matplotlib underneath. A new PR for [Polars](https://github.com/pola-rs/polars/pull/17995) shows how [simple this can be](https://github.com/pola-rs/polars/pull/17995/files#diff-c693ddd3dba49f465b1f2247fa3b5a74c3370067474a4c695d4de7163f97b74bR79-R90) as it's just a very thin wrapper around [Altair](https://altair-viz.github.io/) (disclaimer: I'm a maintainer of Altair).

For duckdb, it would probably be also just a few lines of code to pass through the relevant arguments to Altair. Some of the advantages of going with Altair:
* It should be simple to enable that computations, e.g. for a histogram, are pushed into duckdb thanks to [the DuckDB support in VegaFusion](https://vegafusion.io/duckdb.html): `duckdb_relation.plot.hist(x="column_1", y="column2")`
* Starting with the next release of Altair (5.4), it no longer has any heavy dependencies, e.g. does not depend on pandas, pyarrow, numpy, .... What's left: https://github.com/vega/altair/blob/main/pyproject.toml#L17

CC @jonmmease

**Reply by soerenwolfers** [reply]
@binste pandas' plot interface is a nightmare. You're way better off writing
```
df.pipe(px.line, x='foo', y='bar', ....)
```
which has the additional advantage that it keeps completely separate the dataframe and plotting libraries.( and `.pipe` is useful for other things too)

**Comment by choucavalier** [comment]
hi @binste i'm pretty interested in having a less verbose python expression api directly in duckdb. right now i'm using ibis which adds some overhead (and extra dependencies). i'd rather use duckdb directly. did you make any progress? thanks for this discussion and sharing your ideas and code!

**Reply by binste** [reply]
Same journey for me, Ibis is nice but I'd prefer something without the overhead and extra dependencies :) I think the best "Dataframe-API" candidate for DuckDB right now is the PySpark API. See https://github.com/duckdb/duckdb/discussions/14525. Function-coverage is now at 40% on `main`, with a focus on the most often used ones, compared to ~14% in the latest release. Once [WindowExpression](https://github.com/duckdb/duckdb/discussions/14725#discussioncomment-11200348) is available, one could also easily implement all windowing functions in the PySpark API.

What do you think? I've stopped working on this for now but I think the DuckDB folks are still happy to accept PRs to further extend the PySpark API coverage.

**Reply by MarcoGorelli** [reply]
Hope it's ok for me to interject, I just wanted to say that I've been finding the Python Relational API truly wonderful to build on

DuckDB support in [Narwhals](https://github.com/narwhals-dev/narwhals) is coming along, it's complete enough for 20 out of the 22 tpc-h queries (for the 2 missing ones we need `duckdb.WindowExpression`)

The code example from the original post in Narwhals (>=1.23) becomes:
```python
In [11]: df = nw.from_native(rel)

In [12]: (
    ...:     df
    ...:     .filter(nw.col('j').str.ends_with('e'))
    ...:     .select('i', 'j', two_i = nw.col('i')*2)
    ...:     .sort('i', descending=True)
    ...:     .head(2)
    ...: )
Out[12]:
┌───────────────────────────┐
|    Narwhals LazyFrame     |
|---------------------------|
|┌───────┬─────────┬───────┐|
|│   i   │    j    │ two_i │|
|│ int64 │ varchar │ int64 │|
|├───────┼─────────┼───────┤|
|│     3 │ three   │     6 │|
|│     1 │ one     │     2 │|
|└───────┴─────────┴───────┘|
└───────────────────────────┘
```
which I think is in the spirit of the "# Rewrite with new functionality:" suggested

The main thing we're missing is `duckdb.WindowExpression` - once DuckDB expose that in the Python Relational API, this can become 🔥

Note that support here comes with no extra dependencies, all that's required on top of DuckDB is Narwhals (which itself is pure-Python and with no dependencies of its own). @binste if you fancied trying it out, I'd be curious to see if addresses your needs

**Reply by soerenwolfers** [reply]
@MarcoGorelli I hope you don't mind me asking, I'm genuinely curious and feel I might be missing something: What's the advantage of Narwhals (and Ibis if you have an opinion, I haven't used either) over just converting data to the backend whose syntax you prefer, and then writing your queries/transformations there? I find it hard to imagine a scenario where two conditions are simultaneously true: (1) I absolutely need to have my data stored in a format native to one of the backends whose syntax I don't like as opposed to the one I like or some standard format (e.g. Arrow/Parquet) (2) doing the copy/conversion to the backend I like at the start of my work is prohibitively expensive. Even if such a scenario existed, I don't feel like it's be a great solution to have to learn yet another dataframe API instead of just the one I like plus the one the data is stored in. Of course you might say that I only ever should learn Narwhals, but that feels like an unrealistic ask given that (1) there are competitors of Narwhals that all say they're gonna be the API people should use (2) you'll always get the best performance and support and the least maintenance hassle if you use the API directly provided by the product that does the actual data munging, so not knowing that API is not really an option.

**Reply by MarcoGorelli** [reply]
Thanks @soerenwolfers for asking! This may surprise you, but I agree with everything you wrote! If you're an end-user (say, data scientist, analyst, engineering, etc...) then in general I think your best bet is to pick the right tool for the job and learn to use it really really well.

The use case for Narwhals is for tool-builders, who want to define dataframe logic and then let the user pass in their dataframe of choice. A nice example of this is Plotly, where rewriting their dataframe logic in Narwhals enabled:
- noticeable speedups for Polars users. 3x for most plots, but even >10x in some cases, see https://plotly.com/blog/chart-smarter-not-harder-universal-dataframe-support/
- no loss in functionality (nor extra dependencies) for pandas users. In fact, it had the nice side-effect of fixing some old bugs 🥳 

I would never say that anyone should "only ever learn Narwhals" - in fact, my hope is that if we do our job well, then end users need never care that Narwhals exists, they just keep using their preferred library (pandas, Polars, cuDF, etc...) pass it in to data science tools (plotting, feature engineering, tables, ...), and it can all just work

For eager dataframes, what we're doing seems to be working. I'm hoping to take it further, and enable dataframe-agnostic tools which can also support DuckDB
